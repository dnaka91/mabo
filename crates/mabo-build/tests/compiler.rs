use std::{
    fs,
    path::{Path, PathBuf},
};

use insta::{assert_snapshot, glob, with_settings};
use mabo_build::Opts;
use mabo_parser::Schema;

fn strip_path(path: &Path) -> PathBuf {
    path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs"))
        .or_else(|_| path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs_extra")))
        .unwrap()
        .to_owned()
}

#[test]
fn compile_schema() {
    glob!("inputs/*.mabo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap();
        let value = mabo_compiler::simplify_schema(&value);
        let value = mabo_build::compile_schema(&Opts::default(), &value);
        let value = prettyplease::unparse(&syn::parse2(value.clone()).unwrap());

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("compile", value);
        });
    });
}

#[test]
fn compile_schema_extra() {
    glob!("inputs_extra/*.mabo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap();
        let value = mabo_compiler::simplify_schema(&value);
        let value = mabo_build::compile_schema(&Opts::default(), &value);
        let value = prettyplease::unparse(&syn::parse2(value.clone()).unwrap());

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("compile_extra", value);
        });
    });
}
