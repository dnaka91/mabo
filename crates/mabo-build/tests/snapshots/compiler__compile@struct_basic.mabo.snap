---
source: crates/mabo-build/tests/compiler.rs
description: "/// Basic struct.\nstruct Sample {\n    a: u32 @1,\n    /// Second field\n    b: bool @2,\n}"
input_file: crates/mabo-parser/tests/inputs/struct_basic.mabo
---
#[allow(unused_imports)]
use ::mabo::buf::{Decode, Encode, Size};
/// Basic struct.
#[derive(Clone, Debug, PartialEq)]
#[allow(clippy::module_name_repetitions, clippy::option_option)]
pub struct Sample {
    pub a: u32,
    /// Second field
    pub b: bool,
}
#[automatically_derived]
impl ::mabo::Encode for Sample {
    #[allow(
        clippy::borrow_deref_ref,
        clippy::deref_addrof,
        clippy::explicit_auto_deref,
        clippy::needless_borrow,
        clippy::too_many_lines,
    )]
    fn encode(&self, w: &mut impl ::mabo::BufMut) {
        let Self { a, b } = self;
        ::mabo::buf::encode_field(
            w,
            ::mabo::FieldId::new(1, ::mabo::FieldEncoding::Varint),
            |w| {
                ::mabo::buf::encode_u32(w, *a);
            },
        );
        ::mabo::buf::encode_field(
            w,
            ::mabo::FieldId::new(2, ::mabo::FieldEncoding::Fixed1),
            |w| {
                ::mabo::buf::encode_bool(w, *b);
            },
        );
        ::mabo::buf::encode_u32(w, ::mabo::buf::END_MARKER);
    }
}
#[automatically_derived]
impl ::mabo::Decode for Sample {
    #[allow(clippy::type_complexity, clippy::too_many_lines)]
    fn decode(r: &mut impl ::mabo::Buf) -> ::mabo::buf::Result<Self> {
        let mut a: Option<u32> = None;
        let mut b: Option<bool> = None;
        loop {
            let id = ::mabo::buf::decode_id(r)?;
            match id.value {
                ::mabo::buf::END_MARKER => break,
                1 => a = Some(::mabo::buf::decode_u32(r)?),
                2 => b = Some(::mabo::buf::decode_bool(r)?),
                _ => ::mabo::buf::decode_skip(r, id.encoding)?,
            }
        }
        Ok(Self {
            a: a
                .ok_or(::mabo::buf::Error::MissingField {
                    id: 1,
                    name: Some("a"),
                })?,
            b: b
                .ok_or(::mabo::buf::Error::MissingField {
                    id: 2,
                    name: Some("b"),
                })?,
        })
    }
}
#[automatically_derived]
impl ::mabo::buf::Size for Sample {
    #[allow(
        clippy::borrow_deref_ref,
        clippy::deref_addrof,
        clippy::explicit_auto_deref,
        clippy::needless_borrow,
        clippy::too_many_lines,
    )]
    fn size(&self) -> usize {
        let Self { a, b } = self;
        ::mabo::buf::size_field(1, || { ::mabo::buf::size_u32(*a) })
            + ::mabo::buf::size_field(2, || { ::mabo::buf::size_bool(*b) })
            + ::mabo::buf::size_u32(::mabo::buf::END_MARKER)
    }
}

